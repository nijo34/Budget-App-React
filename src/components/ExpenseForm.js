import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

//big picture goal is to use the local component state to track the changes of all these inputs.
//only when the user submits the form will we actually do something with the information.


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);

        this.state = {
            description :props.expense?props.expense.description :'',
            note: props.expense?props.expense.note:'',
            amount: props.expense?(props.expense.amount).toString():'',
            createdAt : props.expense? moment(props.expense.createdAt):moment(),                                       //current time
            calendarFocused : false,
            error: ''
        }
    }

    onDescriptionChange = (e)=>{                                    //onChange Handlers
        const description = e.target.value                          //populating the state with the new values. 
        this.setState(()=>({description}))
    }

    onNoteChange = (e)=>{
        const note = e.target.value
        this.setState(()=>({note}))
    }

    onAmountChange =(e)=>{
        const amount = e.target.value;
        if(!amount || amount.match( /^\d{1,}(\.\d{0,2})?$/ )){
            this.setState(()=>({amount}))
        }
    }

    onDateChange = (createdAt)=>{
        if(createdAt)                                                   //checks if there is a date specified, disregards the empty date field.
        {
            this.setState(()=>({createdAt}))
        }
        
    }

    onFocusChange = ({focused}) =>{                                     //takes in a destructred object by default whose value is false in the state.
        this.setState(()=>({calendarFocused:focused}))
    }

    onSubmit = (e)=>{
         e.preventDefault();                                            //doesn't go through the full page refresh
        
         if(!this.state.description || !this.state.amount){
             this.setState(()=>({error:'Please provide both description and amount'}))
         }
         else {
             this.setState(()=>({error:''}))
             this.props.onSubmit({
                 description:this.state.description,
                 amount:parseFloat(this.state.amount,10),
                 createdAt: this.state.createdAt.valueOf(),
                 note:this.state.note

             })
         }
        }

    render(){
        return (
                <form 
                className = "form"
                onSubmit= {this.onSubmit} >
                    {this.state.error && <p className = "form__error">{this.state.error}</p>}
                   <input 
                        autoFocus
                        className = "text-input"
                        placeholder = "Description"
                        type="text"
                        value = {this.state.description}
                        onChange = {this.onDescriptionChange}
                    /> 

                    <input 
                        className = "text-input"                        
                        placeholder = "Amount"
                        type = "text"
                        value = {this.state.amount}
                        onChange = {this.onAmountChange}
                    />

                    <SingleDatePicker 
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths ={1}                         //no of months to be shown in the date picker
                        isOutsideRange = {()=>false}                //overriding this prop and setting it to false means that we can go to the dates before the current day as well
                    
                    />

                    <textarea
                        className = "text-area"
                        placeholder = "Add a note for your expenses"
                        value = {this.state.note}
                        onChange = {this.onNoteChange}
                    >
                    </textarea>

                    <div>
                         <button className = "button">Save Expense</button>
                    </div>
                </form>
        )
    } 
}