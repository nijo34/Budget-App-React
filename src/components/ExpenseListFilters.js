import React from 'react';
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import {setTextFilter} from '../actions/filters'
import {sortByDate, sortByAmount,setStartDate,setEndDate} from '../actions/filters'

export class ExpenseListFilters extends React.Component{
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange=(calendarFocused)=>{
        this.setState(()=>({
            calendarFocused
        }))
    }

    onTextChange = (e)=>{                                          //onChange handler takes in a fn which fires every time the text in the input box changes.
        // console.log(e.target.value)                                              // e argument provides us with the new value every time the text changes.
        this.props.setTextFilter(e.target.value)
    }

    onSortChange = (e)=>{
        e.target.value==='amount'?
        this.props.sortByAmount():this.props. sortByDate()
    }

    render(){
        return (
            <div>
                <input type = "text" 
                value = {this.props.filters.text} 
                onChange = {this.onTextChange}/>
        
                <select type = "text" 
                value ={this.props.filters.sortBy} 
                onChange = {this.onSortChange}>
                    <option value ="date">Date</option>
                    <option value="amount">Amount</option>
                </select>   
                
                <DateRangePicker
                    startDate = {this.props.filters.startDate} 
                    endDate = {this.props.filters.endDate}
                    onDatesChange = {this.onDatesChange}
                    focusedInput = {this.state.calendarFocused}
                    onFocusChange = {this.onFocusChange}
                    showClearDates = {true}
                    numberOfMonths ={1}
                    isOutsideRange = {()=>false}
                />
        
            </div>
        )
    }
}


const mapStateToProps = (state)=>{
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch)=>({                        //implicitly returning an object              
    setTextFilter:(text)=>dispatch(setTextFilter(text)),
    sortByDate:()=>dispatch(sortByDate()),
    sortByAmount:()=>dispatch(sortByAmount()),
    setStartDate:(startDate)=>dispatch(setStartDate(startDate)),
    setEndDate:(endDate)=>dispatch(setEndDate(endDate))

})

export default connect(mapStateToProps , mapDispatchToProps)(ExpenseListFilters);