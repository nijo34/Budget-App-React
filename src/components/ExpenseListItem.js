import React from 'react'
import {Link} from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral' 

numeral.register('locale', 'fr', {
    delimiters: {
        thousands: ',',
        decimal: '.'
    },
    abbreviations: {
        thousand: 'k',
        lacs: 'l',
        crore: 'cr'
    },
    ordinal : function (number) {
        return number === 1 ? 'er' : 'Ã¨me';
    },
    currency: {
        symbol: 'Rs '
    }
});

// switch between locales
numeral.locale('fr');

 const ExpenseListItem = ({ id, description , amount , createdAt}) =>(
      <div>
          <Link to = {`/edit/${id}`}><h3>{description}</h3></Link> 
          <p>
          {numeral(amount).format('$0,0.00')} 
          - 
          {moment(createdAt).format('MMM Do, YYYY')} </p>
      </div>
  )


export default ExpenseListItem;