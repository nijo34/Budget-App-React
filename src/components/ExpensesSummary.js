import React from 'react'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expense-total'

const formatter = new Intl.NumberFormat('en', { 
    style: 'currency', 
    currency: 'INR' })


export const ExpensesSummary = ({ expenseCount , expensesTotal })=>{
    const expenseWord = expenseCount === 1? 'expense':'expenses'

    const formattedExpensesTotal = formatter.format(expensesTotal)

    return (
    <div>
        <h1>Viewing { expenseCount } {expenseWord} totalling {formattedExpensesTotal}</h1>
    </div>
)}

const mapStateToProps = (state)=>{ 
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount:visibleExpenses.length,
        expensesTotal:selectExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)