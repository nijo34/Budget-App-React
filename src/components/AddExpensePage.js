import React from 'react'; 
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
    onSubmit = (expense)=>{
        // props.dispatch(addExpense(expense))
        this.props.startAddExpense(expense)
        this.props.history.push('/')                                 //automatically re routes the page to the given path specified within the push method.
    }

    render(){
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit = {this.onSubmit }
        
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({                                                           //this returns an object and on that object, we define various props that are going to call dispatch.
    startAddExpense: (expense) =>  dispatch(startAddExpense(expense))
    })                                                                                                //similar to mapStateToProps, just that it works with dispatch
    
  
export default connect(undefined,mapDispatchToProps)(AddExpensePage);