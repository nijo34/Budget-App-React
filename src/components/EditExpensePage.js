import React from 'react'; 
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense ,removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense)=>{
        this.props.editExpense(this.props.expense.id,expense)
        this.props.history.push('/')
    }

    onClick = () =>{
        this.props.removeExpense({id:this.props.expense.id})
        this.props.history.push('/')
    }

    render (){
        return (
            <div>
            <ExpenseForm
                expense = {this.props.expense}
                onSubmit ={ this.onSubmit}
            />
    
                <button onClick = {this.onClick}>Remove</button>
    
        </div>
        )
    }
}

const mapDispatchToProps = (dispatch,props) => ({
    editExpense:(id,expense) => dispatch(editExpense(id,expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
})

const mapStateToProps = (state, props) =>{                           //have access to the props as the second argument
    return {
        expense : state.expenses.find((expense)=> expense.id === props.match.params.id)  //if it is a match, then it gets assigned to the object 
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);