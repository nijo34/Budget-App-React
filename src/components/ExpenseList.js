import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList =(props) =>(
    <div className = "content-container">
        <div className = "list--header">
        <div className = "show-for-mobile">Expenses</div>
        <div className = "show-for-desktop">Expense</div>
        <div className = "show-for-desktop">Amount</div>
        </div>
        <div className = "list-body">
        {
            props.expenses.length ===0? (
                <div className = "list-item list-item--message">
                    <span>No Expenses</span>
                </div>
            ):
            (
                props.expenses.map((expense)=>{
                    return <ExpenseListItem key ={expense.id} {...expense} />               //using spread operator and passing all the properties within the expenses array
                })
            )
        }
        </div>
    </div>
)

const mapStateToProps = (state)=>{
    return {
        expenses :selectExpenses(state.expenses,state.filters)                      //eariler I passed in the expenses and filters as two seperate key value pairs, but now the filterd list will be shown as the selector is called.
        //expenses:state.expenses,
        //filters:state.filters
    }

}

export default connect(mapStateToProps)(ExpenseList);
