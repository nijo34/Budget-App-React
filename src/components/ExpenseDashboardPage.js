import React from 'react'; 
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from '../components/ExpensesSummary'

const ExpenseDashboardPage = (props) => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />    
        
    </div>
)

export default ExpenseDashboardPage;