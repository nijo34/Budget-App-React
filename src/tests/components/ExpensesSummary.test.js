import React from 'react'
import { shallow } from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'

test('should correctly render expenses summary with 1 expense',()=>{
    const wrapper = shallow(<ExpensesSummary expenseCount = {1} expensesTotal = {234}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should correctly render expenses summary with multiple expenses',()=>{
    const wrapper = shallow(<ExpensesSummary expenseCount = {11} expensesTotal = {21232134}/>)
    expect(wrapper).toMatchSnapshot()
})

