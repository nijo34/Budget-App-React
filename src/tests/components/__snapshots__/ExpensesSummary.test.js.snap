// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expenses summary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    ₹234.00
  </h1>
</div>
`;

exports[`should correctly render expenses summary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    11
     
    expenses
     totalling 
    ₹21,232,134.00
  </h1>
</div>
`;
